source/opt/set_spec_constant_default_value_pass.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "set_spec_constant_default_value_pass.h"
+
+#include <unordered_map>
+
+#include "def_use_manager.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+const uint32_t target_id_operand_index = 0;
+const uint32_t spec_id_operand_index = 2;
+const uint32_t opdecorate_num_operands = 3;
+const uint32_t opspecconstant_inoperand_index = 0;
+};
+
+bool SetSpecConstantDefaultValuePass::Process(ir::Module* module) {
+  bool modified = false;
+  analysis::DefUseManager def_use_mgr(module);
+  // Scan through all the annotation instructions to find 'OpDecorate SpecId'
+  // instructions. Then extract the decoration target id of those instructions
+  // to get the target spec-constant-defining instructions. Replace the default
+  // values (the default bit patterns) with the corresponding values in the
+  // id_to_value_ map.
+  for (ir::Instruction& inst : module->annotations()) {
+    switch (inst.opcode()) {
+      case SpvOp::SpvOpDecorate:
+        if (inst.NumOperands() == opdecorate_num_operands) {
+          uint32_t spec_id = inst.GetSingleWordOperand(spec_id_operand_index);
+          uint32_t target_id =
+              inst.GetSingleWordOperand(target_id_operand_index);
+          ir::Instruction* spec_inst = def_use_mgr.GetDef(target_id);
+          auto iter = spec_id_to_value_.find(spec_id);
+          if (iter != spec_id_to_value_.end()) {
```

Comment:
Break up this method.  The "continue" you use below is a branch to quite a non-local spot.